File: app/models/builders/traffic_lights.rb
# frozen_string_literal: true

class Builders::TrafficLights
  class << self
    def zones_typology
      :single_choice
    end

    def build(retrospective)
      retrospective.add_zone('Support', hint: 'We always get support and help when we ask for it')
      retrospective.add_zone('Team work', hint: 'We are a strong team and collaborate well together')
      retrospective.add_zone('Mission', hint: 'We know why we are here, are aligned with it and excited about it')
      retrospective.add_zone(
        'Codebase',
        hint: 'We are proud of the quality of our code. It is clean, easy to read and has sufficient test coverage'
      )
      retrospective.add_zone('Process', hint: 'Our way of working fits us perfectly')
      retrospective.add_zone('Value', hint: 'We are proud of what we deliver and our end users feel the same')
      retrospective.add_zone('Learning', hint: 'We are learning interesting stuff all the time')
      retrospective.add_zone(
        'Ticket flow',
        hint: 'We get stuff done without useless waiting or delays. Releasing is simple, safe and painless'
      )
      retrospective.add_zone('Fun', hint: 'We love going to work and have great fun working together')
      retrospective.add_zone('Influence', hint: <<~TEXT.squish)
        We are in control of our roadmap, we can intervene in decision process regarding what
        to build and how to build it
      TEXT
    end

    def autovote!(retrospective)
      reflections = sorted_reflections(retrospective)

      count = reflections.count
      votes = []
      now = Time.current
      reflections.each do |reflection|
        count.times do
          votes << reflection.votes.new(
            author: retrospective.facilitator,
            target: reflection,
            content: Reaction::VOTE_EMOJI,
            retrospective: retrospective,
            created_at: now,
            updated_at: now
          )
        end
        count -= 1
      end
      Reaction.insert_all!(votes.map(&:attributes).map(&:compact)) # rubocop:disable Rails/SkipsModelValidations
    end

    def sorted_reflections(retrospective)
      retrospective
        .reflections
        .revealed
        .to_a
        .group_by(&:zone_id)
        .sort_by do |_zone, reflections|
          [
            -reflections.count { |r| r.content == 'red' },
            -reflections.count { |r| r.content == 'orange' },
            -reflections.count { |r| r.content == 'green' }
          ]
        end
        .lazy
        .map(&:last)
        .map(&:first)
    end
  end
end

File: app/javascript/components/retrospectives/TrafficLights.js
import React from 'react'
import { useSelector, shallowEqual } from 'react-redux'
import TrafficLightZone from './traffic_lights/TrafficLightZone'

const TrafficLights = () => {
  const reflections = useSelector(state => state.reflections.ownReflections, shallowEqual)
  const zones = useSelector(state => state.retrospective.zones, shallowEqual)

  const support = zones.find((zone) => zone.name === 'Support')
  const teamWork = zones.find((zone) => zone.name === 'Team work')
  const mission = zones.find((zone) => zone.name === 'Mission')
  const codebase = zones.find((zone) => zone.name === 'Codebase')
  const process = zones.find((zone) => zone.name === 'Process')
  const value = zones.find((zone) => zone.name === 'Value')
  const learning = zones.find((zone) => zone.name === 'Learning')
  const ticketFlow = zones.find((zone) => zone.name === 'Ticket flow')
  const fun = zones.find((zone) => zone.name === 'Fun')
  const influence = zones.find((zone) => zone.name === 'Influence')

  return (
    <div style={{ height: '430px', display: 'flex', flexWrap: 'wrap' }}>
      <TrafficLightZone
        reference={support}
        value={reflections.find((reflection) => reflection.zone.id === support.id)?.content} />

      <TrafficLightZone
        reference={teamWork}
        value={reflections.find((reflection) => reflection.zone.id === teamWork.id)?.content} />

      <TrafficLightZone
        reference={mission}
        value={reflections.find((reflection) => reflection.zone.id === mission.id)?.content} />

      <TrafficLightZone
        reference={codebase}
        value={reflections.find((reflection) => reflection.zone.id === codebase.id)?.content} />

      <TrafficLightZone
        reference={process}
        value={reflections.find((reflection) => reflection.zone.id === process.id)?.content} />

      <TrafficLightZone
        reference={value}
        value={reflections.find((reflection) => reflection.zone.id === value.id)?.content} />

      <TrafficLightZone
        reference={learning}
        value={reflections.find((reflection) => reflection.zone.id === learning.id)?.content} />

      <TrafficLightZone
        reference={ticketFlow}
        value={reflections.find((reflection) => reflection.zone.id === ticketFlow.id)?.content} />

      <TrafficLightZone
        reference={fun}
        value={reflections.find((reflection) => reflection.zone.id === fun.id)?.content} />

      <TrafficLightZone
        reference={influence}
        value={reflections.find((reflection) => reflection.zone.id === influence.id)?.content} />
    </div>
  )
}

export default TrafficLights
